Use the object crate for metadata reading

bjorn3 opened the pull request:
This allows sharing the metadata reader between cg_llvm, cg_clif and other codegen backends.

This is not currently useful for rlib reading with cg_spirv ([rust-gpu](https://github.com/EmbarkStudios/rust-gpu/)) as it uses tar rather than ar as .rlib format, but it is useful for dylib reading required for loading proc macros. (cc @eddyb)

The object crate is already trusted as dependency of libstd through backtrace. As far as I know it supports reading all object file formats used by targets for which we support rust dylibs with crate metadata, but I am not certain. If this happens to not be the case, I could keep using LLVM for reading dylib metadata.

Marked as WIP for a perf run and as it is based on #83637.

rust-highfive commented:
Some changes occured to rustc_codegen_cranelift

cc @bjorn3

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

bjorn3 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit aeee8d87bffdd7370b2c2bf149bf1bfaf7bc18e0 with merge 652060f31a2669f8467ef860c7b663be75b0bd44...
<!-- homu: {"type":"TryBuildStarted","head_sha":"aeee8d87bffdd7370b2c2bf149bf1bfaf7bc18e0","merge_sha":"652060f31a2669f8467ef860c7b663be75b0bd44"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2217318336) [(plain)](https://github.com/rust-lang/rust/commit/aeee8d87bffdd7370b2c2bf149bf1bfaf7bc18e0/checks/2217318336/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
    Checking rustc_resolve v0.0.0 (/checkout/compiler/rustc_resolve)
    Checking rustc_trait_selection v0.0.0 (/checkout/compiler/rustc_trait_selection)
    Checking rustc_codegen_llvm v0.0.0 (/checkout/compiler/rustc_codegen_llvm)
error[E0603]: unit struct `DefaultMetadataLoader` is private
    |
    |
253 |         Box::new(rustc_codegen_ssa::back::metadata::DefaultMetadataLoader)
    |                                                     ^^^^^^^^^^^^^^^^^^^^^ private unit struct
    |
note: the unit struct `DefaultMetadataLoader` is defined here
   --> /checkout/compiler/rustc_codegen_ssa/src/back/metadata.rs:15:1
    |
15  | pub(crate) struct DefaultMetadataLoader;

error: unused import: `crate::llvm`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:1:5
  |
  |
1 | use crate::llvm;
  |     ^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`

error: unused import: `crate::llvm::archive_ro::ArchiveRO`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:2:5
  |
2 | use crate::llvm::archive_ro::ArchiveRO;


error: unused imports: `False`, `ObjectFile`, `mk_section_iter`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:3:19
  |
3 | use crate::llvm::{mk_section_iter, False, ObjectFile};


error: unused import: `rustc_middle::middle::cstore::MetadataLoader`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:4:5
4 | use rustc_middle::middle::cstore::MetadataLoader;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


error: unused import: `rustc_codegen_ssa::METADATA_FILENAME`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:7:5
7 | use rustc_codegen_ssa::METADATA_FILENAME;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unused import: `rustc_data_structures::owning_ref::OwningRef`
error: unused import: `rustc_data_structures::owning_ref::OwningRef`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:8:5
  |
8 | use rustc_data_structures::owning_ref::OwningRef;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unused import: `rustc_data_structures::rustc_erase_owner`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:9:5
9 | use rustc_data_structures::rustc_erase_owner;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unused import: `tracing::debug`
---

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_codegen_ssa" "-p" "rustc_target" "-p" "rustc_incremental" "-p" "rustc_graphviz" "-p" "rustc_apfloat" "-p" "rustc_span" "-p" "rustc_arena" "-p" "rustc_session" "-p" "rustc_lint_defs" "-p" "rustc_feature" "-p" "rustc_serialize" "-p" "rustc_macros" "-p" "rustc_ast" "-p" "rustc_lexer" "-p" "rustc_middle" "-p" "rustc_type_ir" "-p" "rustc_query_system" "-p" "rustc_errors" "-p" "rustc_hir" "-p" "rustc_attr" "-p" "rustc_ast_pretty" "-p" "rustc_index" "-p" "rustc_symbol_mangling" "-p" "rustc_fs_util" "-p" "rustc_data_structures" "-p" "rustc_driver" "-p" "rustc_error_codes" "-p" "rustc_plugin_impl" "-p" "rustc_interface" "-p" "rustc_resolve" "-p" "rustc_ast_passes" "-p" "rustc_traits" "-p" "rustc_infer" "-p" "rustc_privacy" "-p" "rustc_ast_lowering" "-p" "rustc_passes" "-p" "rustc_builtin_macros" "-p" "rustc_parse_format" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_ty_utils" "-p" "rustc_query_impl" "-p" "rustc_expand" "-p" "rustc_trait_selection" "-p" "rustc_mir_build" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_metadata" "-p" "rustc_save_analysis" "-p" "rustc_lint" "-p" "rustc_typeck" "-p" "rustc_hir_pretty" "-p" "rustc_parse" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:35
```

</details>

bjorn3 commented:
@bors r-

bjorn3 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit d2b63d8cd784c90b2c96853d5acea0d6c8ab9976 with merge a7a4d9f098502325534a9e8e22b59dcec45582bc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d2b63d8cd784c90b2c96853d5acea0d6c8ab9976","merge_sha":"a7a4d9f098502325534a9e8e22b59dcec45582bc"} -->

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2217366193) [(plain)](https://github.com/rust-lang/rust/commit/d2b63d8cd784c90b2c96853d5acea0d6c8ab9976/checks/2217366193/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  |
3 | use tracing::debug;
  |     ^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
error: unused import: `rustc_fs_util::path_to_c_string`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:5:5
  |
5 | use rustc_fs_util::path_to_c_string;
---

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_driver" "-p" "rustc_mir_build" "-p" "rustc_apfloat" "-p" "rustc_trait_selection" "-p" "rustc_parse_format" "-p" "rustc_lexer" "-p" "rustc_macros" "-p" "rustc_attr" "-p" "rustc_infer" "-p" "rustc_graphviz" "-p" "rustc_arena" "-p" "rustc_index" "-p" "rustc_hir_pretty" "-p" "rustc_lint" "-p" "rustc_data_structures" "-p" "rustc_save_analysis" "-p" "rustc_plugin_impl" "-p" "rustc_feature" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_parse" "-p" "rustc_middle" "-p" "rustc_query_system" "-p" "rustc_type_ir" "-p" "rustc_typeck" "-p" "rustc_metadata" "-p" "rustc_expand" "-p" "rustc_lint_defs" "-p" "rustc_ast_passes" "-p" "rustc_session" "-p" "rustc_fs_util" "-p" "rustc_ast_pretty" "-p" "rustc_hir" "-p" "rustc_errors" "-p" "rustc_ast" "-p" "rustc_interface" "-p" "rustc_passes" "-p" "rustc_query_impl" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_ast_lowering" "-p" "rustc_symbol_mangling" "-p" "rustc_incremental" "-p" "rustc_traits" "-p" "rustc_resolve" "-p" "rustc_builtin_macros" "-p" "rustc_privacy" "-p" "rustc_ty_utils" "-p" "rustc_target" "-p" "rustc_span" "-p" "rustc_serialize" "-p" "rustc_error_codes" "-p" "rustc_codegen_ssa" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:43
```

</details>

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/2217326658) [(plain)](https://github.com/rust-lang-ci/rust/commit/652060f31a2669f8467ef860c7b663be75b0bd44/checks/2217326658/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_codegen_ssa v0.0.0 (/checkout/compiler/rustc_codegen_ssa)
   Compiling rustc_resolve v0.0.0 (/checkout/compiler/rustc_resolve)
   Compiling rustc_trait_selection v0.0.0 (/checkout/compiler/rustc_trait_selection)
   Compiling rustc_codegen_llvm v0.0.0 (/checkout/compiler/rustc_codegen_llvm)
error[E0603]: unit struct `DefaultMetadataLoader` is private
    |
    |
253 |         Box::new(rustc_codegen_ssa::back::metadata::DefaultMetadataLoader)
    |                                                     ^^^^^^^^^^^^^^^^^^^^^ private unit struct
    |
note: the unit struct `DefaultMetadataLoader` is defined here
error: unused import: `crate::llvm`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:1:5
  |
1 | use crate::llvm;
1 | use crate::llvm;
  |     ^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`

error: unused import: `crate::llvm::archive_ro::ArchiveRO`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:2:5
  |
2 | use crate::llvm::archive_ro::ArchiveRO;


error: unused imports: `False`, `ObjectFile`, `mk_section_iter`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:3:19
  |
3 | use crate::llvm::{mk_section_iter, False, ObjectFile};


error: unused import: `rustc_middle::middle::cstore::MetadataLoader`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:4:5
4 | use rustc_middle::middle::cstore::MetadataLoader;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


error: unused import: `rustc_codegen_ssa::METADATA_FILENAME`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:7:5
7 | use rustc_codegen_ssa::METADATA_FILENAME;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unused import: `rustc_data_structures::owning_ref::OwningRef`
error: unused import: `rustc_data_structures::owning_ref::OwningRef`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:8:5
  |
8 | use rustc_data_structures::owning_ref::OwningRef;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unused import: `rustc_data_structures::rustc_erase_owner`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:9:5
9 | use rustc_data_structures::rustc_erase_owner;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: unused import: `tracing::debug`
---

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "jemalloc llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build --target=x86_64-unknown-linux-gnu --host=x86_64-unknown-linux-gnu --stage 2 library/std --rust-profile-generate=/tmp/rustc-pgo
Build completed unsuccessfully in 0:10:57
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2217414047) [(plain)](https://github.com/rust-lang/rust/commit/570f9a0606a5d1c6250598083b05533ae3035397/checks/2217414047/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  |
3 | use rustc_fs_util::path_to_c_string;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
error: unused import: `std::path::Path`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:4:5
  |
4 | use std::path::Path;
---

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--color" "always" "--features" " llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "-p" "rustc-main" "-p" "rustc_driver" "-p" "rustc_serialize" "-p" "rustc_macros" "-p" "rustc_parse" "-p" "rustc_lexer" "-p" "rustc_mir_build" "-p" "rustc_attr" "-p" "rustc_apfloat" "-p" "rustc_arena" "-p" "rustc_infer" "-p" "rustc_graphviz" "-p" "rustc_index" "-p" "rustc_trait_selection" "-p" "rustc_parse_format" "-p" "rustc_middle" "-p" "rustc_type_ir" "-p" "rustc_query_system" "-p" "rustc_interface" "-p" "rustc_ty_utils" "-p" "rustc_query_impl" "-p" "rustc_resolve" "-p" "rustc_codegen_llvm" "-p" "rustc_llvm" "-p" "rustc_fs_util" "-p" "rustc_symbol_mangling" "-p" "rustc_expand" "-p" "rustc_lint_defs" "-p" "rustc_traits" "-p" "rustc_builtin_macros" "-p" "rustc_ast_passes" "-p" "rustc_incremental" "-p" "rustc_ast_lowering" "-p" "rustc_passes" "-p" "rustc_privacy" "-p" "rustc_feature" "-p" "rustc_mir" "-p" "coverage_test_macros" "-p" "rustc_hir" "-p" "rustc_typeck" "-p" "rustc_target" "-p" "rustc_span" "-p" "rustc_data_structures" "-p" "rustc_save_analysis" "-p" "rustc_ast" "-p" "rustc_session" "-p" "rustc_errors" "-p" "rustc_plugin_impl" "-p" "rustc_ast_pretty" "-p" "rustc_error_codes" "-p" "rustc_lint" "-p" "rustc_hir_pretty" "-p" "rustc_metadata" "-p" "rustc_codegen_ssa" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check
Build completed unsuccessfully in 0:02:22
```

</details>

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/2217380016) [(plain)](https://github.com/rust-lang-ci/rust/commit/a7a4d9f098502325534a9e8e22b59dcec45582bc/checks/2217380016/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  |
3 | use tracing::debug;
  |     ^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
error: unused import: `rustc_fs_util::path_to_c_string`
 --> compiler/rustc_codegen_llvm/src/metadata.rs:5:5
  |
5 | use rustc_fs_util::path_to_c_string;
---

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "jemalloc llvm max_level_info" "--manifest-path" "/checkout/compiler/rustc/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build --target=x86_64-unknown-linux-gnu --host=x86_64-unknown-linux-gnu --stage 2 library/std --rust-profile-generate=/tmp/rustc-pgo
Build completed unsuccessfully in 0:11:08
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2217846718) [(plain)](https://github.com/rust-lang/rust/commit/2e438bafaf782574a0d1674b83b14fd3fb3dc131/checks/2217846718/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Found 435 error codes
Found 0 error codes with no tests
Done!
* 323 features
tidy error: invalid license `Apache-2.0 WITH LLVM-exception` in `wasmparser 0.57.0 (registry+https://github.com/rust-lang/crates.io-index)`
tidy error: Dependencies not explicitly permitted:
* wasmparser 0.57.0 (registry+https://github.com/rust-lang/crates.io-index)
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --stage 2 src/tools/tidy
Build completed unsuccessfully in 0:00:14
```

</details>

bjorn3 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 29f195cc7bb5fbcaa7abfab7a1a9fff63ebcbcb0 with merge dfb3f56d49b7f86c1163b25c8f7d06ca511df6e1...
<!-- homu: {"type":"TryBuildStarted","head_sha":"29f195cc7bb5fbcaa7abfab7a1a9fff63ebcbcb0","merge_sha":"dfb3f56d49b7f86c1163b25c8f7d06ca511df6e1"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2218654226)
Build commit: dfb3f56d49b7f86c1163b25c8f7d06ca511df6e1 (`dfb3f56d49b7f86c1163b25c8f7d06ca511df6e1`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2218654226"},"merge_sha":"dfb3f56d49b7f86c1163b25c8f7d06ca511df6e1"} -->

rust-timer commented:
Queued dfb3f56d49b7f86c1163b25c8f7d06ca511df6e1 with parent 40334da7a6cc0602aa16252cf8f78dc0beb48159, future [comparison URL](https://perf.rust-lang.org/compare.html?start=40334da7a6cc0602aa16252cf8f78dc0beb48159&end=dfb3f56d49b7f86c1163b25c8f7d06ca511df6e1).

rust-timer commented:
Finished benchmarking try commit (dfb3f56d49b7f86c1163b25c8f7d06ca511df6e1): [comparison url](https://perf.rust-lang.org/compare.html?start=40334da7a6cc0602aa16252cf8f78dc0beb48159&end=dfb3f56d49b7f86c1163b25c8f7d06ca511df6e1).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

bjorn3 commented:
#83637 has landed on master, so I rebased it away. In addition I implemented a small optimization that LLVM does to to hopefully fix the perf regression: https://github.com/llvm/llvm-project/blob/9745dce8c3dc6996dae9bab11ff410a12d8daec7/llvm/lib/Support/MemoryBuffer.cpp#L337

bjorn3 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 8fcd03e7358ccb8f568ab3b34f3f31fac4b6ede9 with merge 04773d78a66ab8072a26d0573fbe9eb135c503d5...
<!-- homu: {"type":"TryBuildStarted","head_sha":"8fcd03e7358ccb8f568ab3b34f3f31fac4b6ede9","merge_sha":"04773d78a66ab8072a26d0573fbe9eb135c503d5"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2220375390)
Build commit: 04773d78a66ab8072a26d0573fbe9eb135c503d5 (`04773d78a66ab8072a26d0573fbe9eb135c503d5`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2220375390"},"merge_sha":"04773d78a66ab8072a26d0573fbe9eb135c503d5"} -->

rust-timer commented:
Queued 04773d78a66ab8072a26d0573fbe9eb135c503d5 with parent 3aedcf06b73fc36feeebca3d579e1d2a6c40acc5, future [comparison URL](https://perf.rust-lang.org/compare.html?start=3aedcf06b73fc36feeebca3d579e1d2a6c40acc5&end=04773d78a66ab8072a26d0573fbe9eb135c503d5).

rust-timer commented:
Finished benchmarking try commit (04773d78a66ab8072a26d0573fbe9eb135c503d5): [comparison url](https://perf.rust-lang.org/compare.html?start=3aedcf06b73fc36feeebca3d579e1d2a6c40acc5&end=04773d78a66ab8072a26d0573fbe9eb135c503d5).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf

bjorn3 commented:
That didn't seem to make much of a difference. I don't know if it is the search for the metadata entry or something else that slowed down. @philipc do you think this slowdown is caused by object's archive reader being slower than LLVM's?

philipc commented:
I haven't done any profiling of object's archive reader, but it's simple enough that I don't expect major differences.

Also, you shouldn't need object's wasmparser or flate2 dependencies.

bjorn3 commented:
I don't know what caused the regression and I have no clue how to investigate further. I split a refactoring that should be perf neutral into #83682 and will close this PR.

bjorn3 commented:
Reopened as in https://github.com/rust-lang/rust/pull/84449#discussion_r628300035 @nagisa voiced their support for using object for metadata reading despite the slight performance regression.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2529373226) [(plain)](https://github.com/rust-lang/rust/commit/802fe1756b9fdd6011e17ee6502d6909d2ce55ed/checks/2529373226/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Caused by:
  failed to open: /checkout/Cargo.lock

Caused by:
  Read-only file system (os error 30)
Build completed unsuccessfully in 0:00:31
```

</details>

bjorn3 commented:
Spurious error. Re-running PR checks.

nagisa commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -0,0 +1,65 @@
+    fn get_rlib_metadata(&self, _target: &Target, path: &Path) -> Result<MetadataRef, String> {
+        load_metadata_with(path, |data| {
+            let archive = object::read::archive::ArchiveFile::parse(&*data)
+                .map_err(|e| format!("{:?}", e))?;
```
Why is this error being formatted as a `Debug`, instead of using its `Display` representation?

nagisa commented on compiler/rustc_codegen_ssa/src/back/metadata.rs:
```diff
@@ -0,0 +1,65 @@
+            let archive = object::read::archive::ArchiveFile::parse(&*data)
+                .map_err(|e| format!("{:?}", e))?;
+
+            for entry_result in archive.members() {
```
So, I think this (and use of mmap) may be one plausible reason for the slight performance regression. If my memory of the `ar` format serves me well, obtaining the list of members in an `ar` has to process the file effectively as if it was a ump list.  I suspect that such a read pattern may be a pathological for mmap based I/O: kernel would try loading more data (page?) into memory only for us to inspect the file name and length before we jump to the next entry(-ies), discarding the rest of the data that kernel spent time loading in.

Without digging into LLVM's ArchiveRO implementation I can imagine that more precise reads could be more effective here.

bjorn3 replied:
`ArchiveRO::open` also uses mmap for header reading I think: https://github.com/rust-lang/rust/blob/6e17a5c5fd086ebe6f57216fb3ce5d1d8d6c83e5/compiler/rustc_llvm/llvm-wrapper/ArchiveWrapper.cpp#L66 `MemoryBufferRef` doesn't export any method allowing `read` calls on the mapped file.

nagisa commented on Cargo.lock:
```diff
@@ -2348,8 +2348,10 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
+ "flate2",
  "rustc-std-workspace-alloc",
  "rustc-std-workspace-core",
+ "wasmparser",
```
Hm, given that the other user of this crate is `libstd` (via `gimli`), won't this cause `wasmparser` (and `flate2`) to be compiled into `libstd` too?

bjorn3 replied:
I think I forgot to update `Cargo.lock`. I disabled the wasm feature.

nagisa commented:
The implementation is very reasonable and a huge code quality improvement over the LLVM based version IMO. I'm especially happy with the unification of metadata reading code between backends. The largest regressions introduce a instruction count hit of 0.7% in benchmarks in-line with those I'd expect in terms of the workflows that would be affected by this change most (check, debug).

While the hit is not trivially ignorable, its also small enough, I think, that the maintainability improvements would justify it. And it sounds like there may be a number of low-hanging fruit in archive parsing implementation too.

mati865 commented on compiler/rustc_codegen_llvm/src/base.rs:
```diff
@@ -47,6 +46,21 @@ pub fn write_compressed_metadata<'tcx>(
+    // https://docs.microsoft.com/en-us/windows/win32/debug/pe-format
+    //
+    // As a result, we choose a slightly shorter name! As to why
+    // `.note.rustc` works on MinGW, that's another good question...
```
You can find nice expalination at https://github.com/llvm/llvm-project/blob/7f78e409d0280c62209e1a7dc8c6d1409acc9184/lld/COFF/Writer.cpp#L1194-L1201

bjorn3 replied:
How should I add that link to the source? It is longer than 100 characters and as such tidy would complain.

bjorn3 replied:
That works fine!

nagisa commented:
@bors r+ Thanks.

> I am not sure what you are referring to.

I was referring to the experiment made in [this branch](https://github.com/philipc/object/commits/pr302) (linked to by [one of the comments I linked above](https://github.com/gimli-rs/object/pull/302#issuecomment-836440988)), and in particular [this commit](https://github.com/philipc/object/commit/718be19f9d5414c9f1d5067d64f03c4e8817bc4d). But its fine anyway.

bors commented:
:pushpin: Commit 6381aaf8ae2df01cdb70b6f3123153cf4f1e03cd has been approved by `nagisa`

<!-- @bors r=nagisa 6381aaf8ae2df01cdb70b6f3123153cf4f1e03cd -->
<!-- homu: {"type":"Approved","sha":"6381aaf8ae2df01cdb70b6f3123153cf4f1e03cd","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit 6381aaf8ae2df01cdb70b6f3123153cf4f1e03cd with merge 75da570d784a798a34ff1e5048cd9a6a2fb23170...
<!-- homu: {"type":"BuildStarted","head_sha":"6381aaf8ae2df01cdb70b6f3123153cf4f1e03cd","merge_sha":"75da570d784a798a34ff1e5048cd9a6a2fb23170"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/2585188746)
Approved by: nagisa
Pushing 75da570d784a798a34ff1e5048cd9a6a2fb23170 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/2585188746"},"merge_sha":"75da570d784a798a34ff1e5048cd9a6a2fb23170"} -->
